RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,268 @@
+- `"K"` and `"degC"` have different base unit offsets, they are not compatible.
+- `"s"` and `"ms"` have different base unit scaling, they are not compatible.
+- `"kN"` and `"W.s/mm"` are compatible.
+
+It is a quality of implementation how equality of offsets and scalings are checked.
+
+`empty` units and `undefined` units are always compatible with each other, themselves and any `well-formed` units.
+
+## Unit convertibility
+A well-formed unit `u1` is said to be convertible to another well-formed unit `u2` if:
+* Their base unit factorizations are equal.
+
+Examples:
+- `"N"` is not convertible to `"m/s2"`, since they have different base unit factorizations.
+- `"K"` is convertible to `"degC"`, since they differ by base unit offsets.
+- `"s"` is convertible to `"ms"`, since they differ by base unit scaling.
+- `"kN"` is convertible to `"W.s/mm"`.
+
+`empty` units and `undefined` units are never convertible to each other, themselves or any `well-formed` units.
+
+## Unit inference
+
+### Modified Hindley-Milner
+
+The original Hindley-Milner algorithm is designed to infer types, and consists of building a sequence of substitutions that returns the type of any expression of a given program.
+Applying this algorithm to perform unit inference in Modelica comes with two important differences:
+  * Not all variables must have a unit in a given model
+  * There are operation on units, that is, not all constraints are of the form  _`var1.unit` is compatible with `var2.unit`_.
+
+The following describes a modified version of the Hindley-Milner's algorithm to perform unit inference in the Modelica language.
+
+* Traverse the model and collect constraints before any evaluations has occured.
+* After evaluable parameters have been evaluated, evaluate both sides of the constraints. (This turns constant Integer expressions into literals, evaluates certain `if-condition`s and propagates `empty` and `undefined` units to the top of the constraints.)
+* Discard any constraints of the form _`u` must be compatible with empty unit_, or _`u` must be compatible with undefined unit_, as they are trivially satisfied.
+* For all constraints:
+    * If the constaint is of the form _`well-formed` unit `u1` must be compatible with `well-formed` unit `u2`_, check that this is the case, it is an error otherwise.
+    * Otherwise:
+        * Rewrite the constraint to be of the form `var.unit` is compatible with `u` where `u` is a unit meta-expression that doesn't contain references to the unit variable `var.unit`, if it is not possible, discard the constraint.
+        * Append the substitution _`var.unit` => `u`_ to the sequence of substitutions and apply it to all remaining constraints.
+
+* For every variable `var` without a unit-attribute, apply the sequence of substitutions to `var.unit`.
+    * If it results in a `well-formed` unit, then this is the unit of `var`.
+    * The unit of `var` is left unknown, otherwise
```

Comment:
Complete sentence (no need for _left_, as it is now that all variables get their unit, as far as I understand):
```suggestion
    * Otherwise, the unit of `var` is unknown.
```